packages/pigeon/example/app/ios/Runner/Messages.g.swift

```diff
@@ -222,6 +223,24 @@ class ExampleHostApiSetup {
     } else {
       sendMessageChannel.setMessageHandler(nil)
     }
+    let sendMessageModernAsyncChannel = FlutterBasicMessageChannel(
+      name:
+        "dev.flutter.pigeon.pigeon_example_package.ExampleHostApi.sendMessageModernAsync\(channelSuffix)",
+      binaryMessenger: binaryMessenger, codec: codec)
+    if let api = api {
+      sendMessageModernAsyncChannel.setMessageHandler { message, reply in
+        let args = message as! [Any?]
+        let messageArg = args[0] as! MessageData
+        Task {
+          let result = await api.sendMessageModernAsync(message: messageArg)
+          await MainActor.run {
```

Comment:
Why doesn't the old version switch thread? Is `setMessageHandler`'s callback guaranteed to be called on main thread?

Reply:
The old version uses the callback approach that doesn't introduce asynchronous things like task so there was no need to switch to main to reply via channel

Reply:
> Is `setMessageHandler`'s callback guaranteed to be called on main thread? 

I guess so as it is native platform channel way of communication

Reply:
After marking the func definition as `@MainActor` [discussed here](https://github.com/flutter/packages/pull/8341/files#r1898069753), here we can just do: 

```
Task { @MainActor in
  let result = await api.send...
  reply(wrapResult(result))
}
```

Reply:
> After marking the func definition as `@MainActor` [discussed here](https://github.com/flutter/packages/pull/8341/files#r1898069753), here we can just do: 
> 
> ```
> Task { @MainActor in
>   let result = await api.send...
>   reply(wrapResult(result))
> }
> ```
> 

No problem it could be easily done.

But could you explain to me how it would work?

If we mark the whole function with @MainActor doesn't that mean that some expensive async/await function call could potentially be called at main?

Reply:
> If we mark the whole function with @mainactor doesn't that mean that some expensive async/await function call could potentially be called at main?

The API contract is that this function will be called on main. The old API is also called on main and this does not change that. Marking it as `@MainActor` will allow compiler to enforce the implementor of this API to only use API that's safe to be used on main thread. That's one of the benefit of swift concurrency. 

An expensive aync/await call inside this function will be called on the thread specified by that call. For example, 
```
@MainActor
func sendMessageModernAsync() {
  let foo = await myActor.getExpensiveFoo()
}
```

Here `getExpensiveFoo` will be called under `myActor`'s async context, rather than that of the main actor.

Reply:
Done